File: examples/record-browser-session/frameworks/playwright/csharp/ScreenRecording.cs
// Records a browser session as a .webm file using Browserless CDP commands.
// Must reuse the existing context and page from ConnectOverCDPAsync — creating new ones
// starts a fresh session that isn't wired to the recording.
//
// Install: dotnet add package Microsoft.Playwright && playwright install chromium
// Run:     dotnet run

using Microsoft.Playwright;
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;

class ScreenRecording
{
    public static async Task Main()
    {
        const string TOKEN = "YOUR_API_TOKEN_HERE";
        string WS_ENDPOINT = $"wss://production-sfo.browserless.io?token={TOKEN}&headless=false&stealth&record=true";

        using var playwright = await Playwright.CreateAsync();
        var browser = await playwright.Chromium.ConnectOverCDPAsync(WS_ENDPOINT);

        // Reuse the existing context and page — new ones won't be wired to the recording.
        var context = browser.Contexts[0];
        var page = context.Pages[0];

        // Set viewport before starting — dimensions are fixed for the entire recording.
        await page.SetViewportSizeAsync(1280, 720);

        var cdpSession = await context.NewCDPSessionAsync(page);
        await cdpSession.SendAsync("Browserless.startRecording");

        await page.GotoAsync("https://example.com");
        await Task.Delay(5000);

        await page.GotoAsync("https://example.com/about");
        await Task.Delay(5000);

        // base64 encoding is required — CDP can't transfer raw binary over its text protocol.
        var response = await cdpSession.SendAsync(
            "Browserless.stopRecording",
            new Dictionary<string, object> { ["encoding"] = "base64" }
        );
        byte[] data = Convert.FromBase64String(response.Value<string>("value"));
        await File.WriteAllBytesAsync("recording.webm", data);

        Console.WriteLine("Recording saved to recording.webm");
        await browser.CloseAsync();
    }
}

File: examples/record-browser-session/frameworks/playwright/java/ScreenRecording.java
// Records a browser session as a .webm file using Browserless CDP commands.
// Must reuse the existing context and page from connectOverCDP — creating new ones
// starts a fresh session that isn't wired to the recording.
//
// Add to pom.xml:
//   <dependency>
//     <groupId>com.microsoft.playwright</groupId>
//     <artifactId>playwright</artifactId>
//     <version>1.44.0</version>
//   </dependency>
//
// Run: mvn exec:java

import com.microsoft.playwright.*;
import java.nio.file.*;
import java.util.Base64;
import java.util.Map;

public class ScreenRecording {
    public static void main(String[] args) throws Exception {
        String TOKEN = "YOUR_API_TOKEN_HERE";
        String WS_ENDPOINT = String.format(
            "wss://production-sfo.browserless.io?token=%s&headless=false&stealth&record=true",
            TOKEN
        );

        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(WS_ENDPOINT);
            try {
                // Reuse the existing context and page — new ones won't be wired to the recording.
                BrowserContext context = browser.contexts().get(0);
                Page page = context.pages().get(0);

                // Set viewport before starting — dimensions are fixed for the entire recording.
                page.setViewportSize(1280, 720);

                CDPSession cdpSession = context.newCDPSession(page);
                cdpSession.send("Browserless.startRecording", null);

                page.navigate("https://example.com");
                Thread.sleep(5000);

                page.navigate("https://example.com/about");
                Thread.sleep(5000);

                // base64 encoding is required — CDP can't transfer raw binary over its text protocol.
                var response = cdpSession.send(
                    "Browserless.stopRecording",
                    Map.of("encoding", "base64")
                );
                var valueElement = response.get("value");
                if (valueElement == null) throw new IllegalStateException("stopRecording response missing 'value'");
                byte[] data = Base64.getDecoder().decode(valueElement.getAsString());
                Files.write(Paths.get("recording.webm"), data);

                System.out.println("Recording saved to recording.webm");
            } finally {
                browser.close();
            }
        }
    }
}
